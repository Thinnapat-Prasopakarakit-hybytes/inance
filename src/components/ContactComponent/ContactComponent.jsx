import "./ContactComponent.scss";
import { Map } from "@vis.gl/react-google-maps";
import { useIntl } from "react-intl";
import { logEvent as logEventGA } from "../../analytics";
import { z } from "zod";
import { collection, addDoc } from "firebase/firestore";
import { logEvent } from "firebase/analytics";
import { db, analytics } from "../../firebase/config";
import { useState } from "react";
import { v4 as uuidv4 } from "uuid";

const ContactComponent = () => {
  const { messages, locale } = useIntl();
  const [formData, setFormData] = useState({
    name: "",
    phone: "",
    email: "",
    message: "",
    honeypot: "",
  });
  const [errors, setErrors] = useState({});
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [submitStatus, setSubmitStatus] = useState(null);

  const ContactSchema = z.object({
    name: z.string().min(3, { message: messages.contact.errors.name }),
    phone: z.string().min(10, { message: messages.contact.errors.phone }),
    email: z
      .string()
      .min(1, { message: messages.contact.errors.email.required })
      .email({ message: messages.contact.errors.email.invalid }),
    message: z.string().min(10, { message: messages.contact.errors.message }),
  });

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((prev) => ({
      ...prev,
      [name]: value,
    }));
    if (errors[name]) {
      setErrors((prev) => ({
        ...prev,
        [name]: "",
      }));
    }
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    logEventGA("Form", "Submit", "Contact Form");
    logEvent(analytics, "submit_form", {
      form_name: "Contact Us",
      success: true,
    });
    setIsSubmitting(true);
    setSubmitStatus(null);

    if (formData.honeypot) {
      setFormData({
        name: "",
        phone: "",
        email: "",
        message: "",
        honeypot: "",
      });
      setSubmitStatus({
        type: "success",
        message: messages.contact.status.success,
      });
      return;
    }

    try {
      ContactSchema.parse(formData);
      setErrors({});

      const { honeypot, ...submissionData } = formData;

      await addDoc(collection(db, "contacts"), {
        ...submissionData,
        id: uuidv4(),
        createdAt: new Date().toISOString(),
        language: locale,
      });

      setFormData({
        name: "",
        phone: "",
        email: "",
        message: "",
        honeypot: "",
      });
      setSubmitStatus({
        type: "success",
        message: messages.contact.status.success,
      });
    } catch (error) {
      if (error instanceof z.ZodError) {
        const formattedErrors = error.errors.reduce((acc, err) => {
          acc[err.path[0]] = err.message;
          return acc;
        }, {});
        setErrors(formattedErrors);
      } else {
        setSubmitStatus({
          type: "error",
          message: messages.contact.status.error,
        });
        console.error("Error submitting form:", error);
      }
    } finally {
      setIsSubmitting(false);
    }
  };

  return (
    <section className="contact_section layout_padding">
      <div className="container">
        <div className="heading_container">
          <h2>{messages.contact.title}</h2>
        </div>
        <div className="row">
          <div className="col-md-6">
            <form onSubmit={handleSubmit}>
              <div className="honeypot-field">
                <input
                  type="text"
                  name="honeypot"
                  value={formData.honeypot}
                  onChange={handleChange}
                  style={{ display: "none" }}
                />
              </div>
              <div>
                <input
                  type="text"
                  name="name"
                  placeholder={messages.contact.name}
                  value={formData.name}
                  onChange={handleChange}
                  className={errors.name ? "error" : ""}
                />
                {errors.name && (
                  <span className="error-message">{errors.name}</span>
                )}
              </div>
              <div>
                <input
                  type="phone"
                  name="phone"
                  placeholder={messages.contact.phone}
                  value={formData.phone}
                  onChange={handleChange}
                  className={errors.phone ? "error" : ""}
                />
                {errors.phone && (
                  <span className="error-message">{errors.phone}</span>
                )}
              </div>
              <div>
                <input
                  type="text"
                  name="email"
                  placeholder={messages.contact.email}
                  value={formData.email}
                  onChange={handleChange}
                  className={errors.email ? "error" : ""}
                />
                {errors.email && (
                  <span className="error-message">{errors.email}</span>
                )}
              </div>
              <div>
                <textarea
                  type="text"
                  name="message"
                  className={`message-box ${errors.message ? "error" : ""}`}
                  placeholder={messages.contact.message}
                  value={formData.message}
                  onChange={handleChange}
                />
                {errors.message && (
                  <span className="error-message">{errors.message}</span>
                )}
              </div>
              <div className="d-flex">
                <button type="submit" disabled={isSubmitting}>
                  {isSubmitting ? "SENDING..." : messages.contact.send}
                </button>
              </div>
              {submitStatus && (
                <div className={`submit-status ${submitStatus.type}`}>
                  {submitStatus.message}
                </div>
              )}
            </form>
          </div>
          <div className="col-md-6">
            <div className="map_container">
              <Map
                defaultZoom={18}
                defaultCenter={{
                  lat: 40.712775,
                  lng: -74.005973,
                }}
                gestureHandling={"greedy"}
                disableDefaultUI
              />
            </div>
          </div>
        </div>
      </div>
    </section>
  );
};

export default ContactComponent;
