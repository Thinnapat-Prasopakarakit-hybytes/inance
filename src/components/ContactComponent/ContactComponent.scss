@use "../../styles/partials/variables" as *;
@use "../../styles/partials/mixins" as *;

.heading_container {
  margin-bottom: 30px;
}

.contact_section {
  form {
    @include tablet {
      margin-bottom: 45px;
    }
  }

  input,
  textarea {
    width: 100%;
    border: none;
    height: 50px;
    padding: 0 25px;
    background-color: $white;
    outline: none;
    color: $black;
    margin-bottom: 25px;
    box-shadow: 0 0 7px 0 rgba($color: $black, $alpha: 0.2);

    &::placeholder {
      color: #737272;
    }

    &.message-box {
      height: 135px;
    }

    &.error {
      margin-bottom: 0;
      border-color: #dc3545;
      background-color: #fff8f8;
    }
  }

  textarea {
    height: 135px;
    padding: 15px 25px;
    resize: none;
  }

  .error-message {
    color: #dc3545;
    font-size: 0.875rem;
    margin-top: 0.25rem;
    display: block;
    margin-bottom: 25px;
  }

  .submit-status {
    margin-top: 1rem;
    padding: 0.75rem;
    border-radius: 4px;
    text-align: center;

    &.success {
      background-color: #d4edda;
      color: #155724;
      border: 1px solid #c3e6cb;
    }

    &.error {
      background-color: #f8d7da;
      color: #721c24;
      border: 1px solid #f5c6cb;
    }
  }

  button {
    border: none;
    @include hero_btn($primary1, $white, 10px, 55px, 0);
    margin-top: 15px;

    &:disabled {
      opacity: 0.7;
      cursor: not-allowed;
    }
  }

  .map_container {
    height: 360px;

    .map {
      height: 100%;
    }
  }
}
