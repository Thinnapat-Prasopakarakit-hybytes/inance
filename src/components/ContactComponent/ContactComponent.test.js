import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import ContactComponent from "./ContactComponent";
import { APIProvider } from "@vis.gl/react-google-maps";

jest.mock("@vis.gl/react-google-maps", () => ({
  APIProvider: ({ children }) => (
    <div data-testid="mock-api-provider">{children}</div>
  ),
  Map: ({ defaultZoom, defaultCenter, gestureHandling, disableDefaultUI }) => (
    <div
      data-testid="mock-map"
      data-props={JSON.stringify({
        defaultZoom,
        defaultCenter,
        gestureHandling,
        disableDefaultUI,
      })}
    />
  ),
}));

const renderContactComponent = () => {
  return render(
    <BrowserRouter>
      <APIProvider apiKey={process.env.REACT_APP_GOOGLE_MAPS_API_KEY}>
        <ContactComponent />
      </APIProvider>
    </BrowserRouter>
  );
};

describe("ContactComponent", () => {
  it("should renders contact form", () => {
    renderContactComponent();
    expect(screen.getByText("Contact Us")).toBeInTheDocument();
  });

  it("should render with all content", () => {
    renderContactComponent();
    expect(screen.getByText("Contact Us")).toBeInTheDocument();

    expect(screen.getByPlaceholderText("Name")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("Phone Number")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("Email")).toBeInTheDocument();
    expect(screen.getByPlaceholderText("Message")).toBeInTheDocument();

    expect(screen.getByText("SEND")).toBeInTheDocument();
  });

  it("should render all input fields", () => {
    renderContactComponent();
    const nameInput = screen.getByPlaceholderText("Name");
    const phoneInput = screen.getByPlaceholderText("Phone Number");
    const emailInput = screen.getByPlaceholderText("Email");
    const messageInput = screen.getByPlaceholderText("Message");

    fireEvent.change(nameInput, { target: { value: "John Doe" } });
    expect(nameInput.value).toBe("John Doe");

    fireEvent.change(phoneInput, { target: { value: "12345678" } });
    expect(phoneInput.value).toBe("12345678");

    fireEvent.change(emailInput, { target: { value: "john.doe@example.com" } });
    expect(emailInput.value).toBe("john.doe@example.com");

    fireEvent.change(messageInput, { target: { value: "Test Message" } });
    expect(messageInput.value).toBe("Test Message");
  });

  it("should simulate form submission", () => {
    renderContactComponent();
    const sendButton = screen.getByText("SEND");

    fireEvent.click(sendButton);
    expect(sendButton).toBeEnabled();
  });

  it("should render map with correct props", () => {
    renderContactComponent();
    const mapElement = screen.getByTestId("mock-map");
    const mapProps = JSON.parse(mapElement.dataset.props);

    expect(mapProps).toEqual({
      defaultZoom: 18,
      defaultCenter: {
        lat: 40.712775,
        lng: -74.005973,
      },
      gestureHandling: "greedy",
      disableDefaultUI: true,
    });
  });

  it("should match snapshot", () => {
    const { asFragment } = renderContactComponent();
    expect(asFragment()).toMatchSnapshot();
  });
});
