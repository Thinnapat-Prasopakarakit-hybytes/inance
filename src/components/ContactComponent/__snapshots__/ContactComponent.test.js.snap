// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactComponent should match snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="mock-api-provider"
  >
    <section
      class="contact_section layout_padding"
    >
      <div
        class="container"
      >
        <div
          class="heading_container"
        >
          <h2>
            Contact Us
          </h2>
        </div>
        <div
          class="row"
        >
          <div
            class="col-md-6"
          >
            <form
              action=""
            >
              <div>
                <input
                  placeholder="Name"
                  type="text"
                />
              </div>
              <div>
                <input
                  placeholder="Phone Number"
                  type="text"
                />
              </div>
              <div>
                <input
                  placeholder="Email"
                  type="email"
                />
              </div>
              <div>
                <input
                  class="message-box"
                  placeholder="Message"
                  type="text"
                />
              </div>
              <div
                class="d-flex"
              >
                <button>
                  SEND
                </button>
              </div>
            </form>
          </div>
          <div
            class="col-md-6"
          >
            <div
              class="map_container"
            >
              <div
                data-props="{\\"defaultZoom\\":18,\\"defaultCenter\\":{\\"lat\\":40.712775,\\"lng\\":-74.005973},\\"gestureHandling\\":\\"greedy\\",\\"disableDefaultUI\\":true}"
                data-testid="mock-map"
              />
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
